unliftio-core/src/Control/Monad/IO/Unlift.hs

```diff
@@ -41,13 +41,15 @@ newtype UnliftIO m = UnliftIO { unliftIO :: forall a. m a -> IO a }
 --
 -- * @unliftIO u (m >>= f) = unliftIO u m >>= unliftIO u . f@
 --
--- The third is a currently nameless law which ensures that the
--- current context is preserved.
+-- Instances of @MonadUnliftIO@ must also satisfy these idempotency laws:
 --
--- * @askUnliftIO >>= (\u -> liftIO (unliftIO u m)) = m@
+-- * @askUnliftIO >>= \\_ -> m = m@
 --
--- If you have a name for this, please submit it in a pull request for
--- great glory.
+-- * @liftIO . unliftIO u = id@
```

Comment:
This doesn't seem correct. I can capture a `u` from a different monadic context and then `liftIO . unliftIO u` would in fact be different from `id`.

Reply:
hm. What's an example where `liftIO . unliftIO u` is not `id`?

And is `askUnliftIO >>= (\u -> liftIO (unliftIO u m1)  >> liftIO (unliftIO u m2))` equivalent to `m1 >> m2`? Or is it necessary that `u` is only consumed once for this law?

Reply:
oh are you talking about the case of
```haskell
u <- local (+ 1) $ askUnliftIO
ask -- 1
liftIO (unliftIO u ask) -- 2
```

In other words, you're saying that it's only necessarily `id` if run within the same monadic context as `askUnliftIO` was called in?